ee/modules/600-flant-integration/images/flant-pricing/hooks/controller_metrics.py

```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python3
+#
+# Copyright 2023 Flant JSC
+# Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+#
+# This hook is responsible for generating metrics for d8 controllers resource consumption.
+
+
+from utils import prometheus_query_value, prometheus_metric_builder, prometheus_function_builder, check_for_generate_mock
+from typing import List, Dict, Any, Optional
+from dataclasses import dataclass
+from shell_operator import hook
+from pathlib import Path
+import json
+import os
+
+
+RESOURCE_CONSUMPTION_QUERY = '''
+sum (
+    ( {} )
+    + on(pod) group_left(controller_name, controller_type)
+    ( {} * 0 )
+)
+'''
+PROMETHEUS_INTERVAL = "5m"
+
+
+PROMETHEUS_METRIC_GROUP = "group_d8_controller_metrics"
+PROMETHEUS_CPU_METRIC_NAME = "flant_pricing_controller_average_cpu_usage_seconds"
+PROMETHEUS_MEMORY_METRIC_NAME = "flant_pricing_controller_average_memory_working_set_bytes:without_kmem"
+
+MOCK_DATA_FILE = None
+if check_for_generate_mock():
+    MOCK_DATA_FILE = os.path.join(os.path.dirname(__file__), f"{Path(__file__).stem}_mock_data.json")
+    with open(MOCK_DATA_FILE, "w") as f:
+        json.dump([], f)
+
+
+@dataclass
+class Controller:
+    name: str
+    namespace: str
+    kind: str
+    module: str
+    cpu: Optional[float] = None
+    memory: Optional[float] = None
+
+
+def main(ctx: hook.Context):
+    # Generate list of Controllers from snapshots
+    controllers = process_controllers(snapshots=ctx.snapshots)
+
+    # Generate metrics from Controllers list
+    metrics = generate_metrics(controllers=controllers)
```

Comment:
Nitpicking. I would not use keyword arguments when a function has very few arguments. The name of variables is clean enough to see what's going on (ctx.snapshots, controllers)